### PR TITLE
Add fenvol-puzzle.lic: Fenvol's Library puzzle solver

### PR DESCRIPTION
**Pull request by MahtraDR**
## Summary
- Merges two existing puzzle scripts (`fenvol-puzzle` and `puzzle-shelf`) into one comprehensive Fenvol's Library puzzle solver
- Uses poem-based item matching (no hardcoded book noun list to maintain), DFS exploration via `XMLData.room_title`, adjective-aware container detection with open-fallback, and `Flags`-based completion detection
- Includes library entry flow with card redemption, configurable repeat modes (`N` runs or `repeat` until out of cards), and YAML settings for reward stowing/discarding

## Features
- **Entry flow**: touches door, redeems library card, handles butler gate and confirmation prompts
- **Poem matching**: reads notecard, extracts poem between delimiters, normalizes text, strips articles for flexible substring matching
- **Container detection**: scans room descriptions for container nouns with up to 2 adjectives, tries opening with fallback from most to least specific
- **DFS exploration**: visits each room once using `XMLData.room_title` for deduplication, backtracks via `OPPOSITES` mapping
- **Reward handling**: stows into configured container or discards nouns on the discard list
- **Repeat modes**: single run, exact N runs, or infinite until out of library cards

## YAML Settings
```yaml
fenvol_container: canvas sack in my back   # optional -- stow reward here
fenvol_discard_list:                        # optional -- nouns to discard
  - dress
  - bodice
```

## Test plan
- [x] 136 RSpec examples, 0 failures
- [x] Rubocop clean (0 offenses)
- [ ] Manual test in-game at Fenvol's Library with single run
- [ ] Manual test with `repeat` mode across multiple library cards
- [ ] Verify reward stowing into configured container
- [ ] Verify discard list correctly puts items in bucket

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Fenvol's Library puzzle automation: run once, fixed-repeat, or continuous; validates empty hands, automates entry, navigation, poem-based item matching/turning, and configurable stow/discard handling for rewards.
* **Tests**
  * Comprehensive test suite with a harness covering text/XML normalization, container/opening candidate logic, room traversal/solve flows, repeat/stop behavior, reward handling and edge cases.
* **Chore**
  * Clears puzzle completion and reward state on shutdown.

[Review Change Stack](https://app.coderabbit.ai/change-stack/elanthia-online/dr-scripts/pull/7403?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by MahtraDR** [comment]
@coderabbitai All three findings addressed:

1. **Misleading `in_poem?('grim')` test name** -- fixed in 4ef6998 (renamed to "matches partial words via substring")
2. **Misleading `in_poem?('')` test name** -- fixed in 4ef6998 (renamed to "returns true for empty item text")
3. **`try_open_candidates('')` returns `[nil]`** -- fixed in 6103508 with early-return guard (`return [] if words.empty?`), specs updated to expect `[]`

**Comment by MahtraDR** [comment]
@coderabbitai review this PR and all commits in full.

**Comment by MahtraDR** [comment]
@coderabbitai Addressed in ea3e2b4f:

**Major (valid fix):** `try_open_candidates` now includes the full noun phrase as the first candidate. For "ichorous green ottoman", candidates are now `["ichorous green ottoman", "green ottoman", "ichorous ottoman", "ottoman"]`. The `.uniq` call handles single-word deduplication.

**Three duplicate comments:** The test name fixes for `in_poem?('grim')` and `in_poem?('')` were already resolved in 4ef6998 (prior review cycle).

File: fenvol-puzzle.lic [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+    words = noun_phrase.split
+    return [] if words.empty?
+
+    noun = words.last
+    adjectives = words[0..-2]
+
+    candidates = adjectives.reverse.map { |adj| "#{adj} #{noun}" }
+    candidates << noun
+    candidates.uniq
+  end
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply fenvol-puzzle.lic:403]: @coderabbitai Fixed in ea3e2b4f. Full noun phrase is now the first candidate, falling back through individual adjectives. `.uniq` handles single-word deduplication.

[the remark this answers is not in the document]
Comment (MahtraDR) [reply fenvol-puzzle.lic:403]: @coderabbitai Reverted in 0ba4d589. DR's game parser only accepts adjective+noun pairs, not multi-word phrases. `open ichorous green ottoman` would always fail -- the game needs `open green ottoman` or `open ichorous ottoman`. The full phrase as first candidate would waste a command on every multi-adjective container. Added a YARD comment documenting this parser constraint.

File: fenvol-puzzle.lic [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+    return true unless item && !item.to_s.empty?
+
+    noun = item.to_s.downcase.split.last
+
+    if @discard_list.include?(noun)
+      echo "Discarding #{item} (on discard list)."
+      return DRCI.put_away_item_unsafe?("my #{noun}", 'bucket')
+    end
+
+    return true if @containers.empty?
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply fenvol-puzzle.lic:465]: @coderabbitai Intentional behavior, not a bug. When `@containers` is empty, the user chose not to configure `fenvol_container` -- they want the reward left in hand. Returning `false` would break single-run users with a misleading "Container may be full" error when no container is configured. For repeat mode, `DRCI.stow_hands` in `redeem_card` tries the default game container as a first pass, and `stow_reward` is called as fallback after that. If neither works, the user needs to configure `fenvol_container`.
